# 01 Data preparation

Processing xlsx files from data folder into suitable inputs and generate other input files

In [1]:
import pandas as pd
import numpy as np
import os
import pickle

In [2]:
cell_line ='BC3C'
data_dir = f"/home/jing/Phd_project/project_UCD_blca/blca_publication_OUTPUT/blca_publication_OUTPUT_LINCS/00_outputs_2020_{cell_line}/"
info_dir = data_dir
out_dir = f"/home/jing/Phd_project/project_UCD_blca/blca_publication_OUTPUT/blca_publication_OUTPUT_bmra/blca_publication_OUTPUT_bmra_{cell_line}/00_outputs_2020_{cell_line}"

#os.makedirs(info_dir, exist_ok = True)
os.makedirs(out_dir, exist_ok = True)

## Modules

Load data about modules and drugs.

In [4]:
### DATA
### remove whitespaces in names (modules), remove duplicates

modules_df = pd.read_excel(
    os.path.join(info_dir, "ALL_DATA_2020_Jing.xlsx"), sheet_name = "modules", index_col = 0)
# display(modules_df)

selected_modules = modules_df.index.tolist()
print(len(selected_modules), ' - Size after reading')

# remove duplicates
selected_modules = modules_df.index.unique().tolist()
print(len(selected_modules), ' - Size after remove duplicates')

# remove whitespaces in modules' names 
selected_modules = [d.strip() for d in selected_modules]

print('Selected modules list: ', len(selected_modules), selected_modules)


10  - Size after reading
10  - Size after remove duplicates
Selected modules list:  10 ['CDK1', 'CDK2', 'CDK4_6', 'EGFR', 'PI3K', 'FGFR', 'TOP2A', 'p53', 'Estrogen', 'Androgen']


In [5]:
### DATA
### remove whitespaces in names (modules, drugs), remove duplicates
### check the dimensions of the indicator IC50 1 uM = 1000 nM
### copy-paste as values, numbers, no formulas

IC50_df = pd.read_excel(
    os.path.join(info_dir, "ALL_DATA_2020_Jing.xlsx"), sheet_name = "IC50s")
IC50_df.drop(columns=['Unnamed: 3','Unnamed: 4'],inplace=True)

print(len(IC50_df.index), ' - Size after reading')
# display(IC50_df)

# rename
IC50_df = IC50_df.rename(columns = {"IC50, uM": "IC50"})

# manually correcting value IC50, Example  for IOX2 -> 30 nM  
#IC50_df.loc[IC50_df.index == 'IOX2', IC50_df.columns == 'IC50'] = 30/1000

# remove non-selected modules, modules' names with whitespaces or empty
IC50_df = IC50_df[IC50_df.Module.isin(selected_modules)]

print(len(IC50_df.index), ' - Size after remove modules')
# display(IC50_df)

# remove duplicates 
# Considering certain columns is optional. 
# Indexes, including time indexes are ignored.
IC50_df = IC50_df.drop_duplicates()

print(len(IC50_df.index), ' - Size after remove duplicates')
display(IC50_df)

39  - Size after reading
39  - Size after remove modules
39  - Size after remove duplicates


Drug    Module       IC50
0           flufenamic-acid  Androgen    3.00000
1                nandrolone  Androgen    9.00000
2               oxandrolone  Androgen  190.30000
3    testosterone-enanthate  Androgen  200.00000
4   testosterone-propionate  Androgen  124.00000
5               JNJ-7706621      CDK1    0.02700
6                PHA-793887      CDK1    0.18000
7               roscovitine      CDK2    2.00000
8                 alvocidib    CDK4_6    0.12000
9               palbociclib    CDK4_6    0.04500
10                 afatinib      EGFR    0.03000
11                erlotinib      EGFR    0.00600
12                gefitinib      EGFR    0.17100
13                lapatinib      EGFR    0.03240
14               vandetanib      EGFR    1.50000
15               dienestrol  Estrogen  200.00000
16      estradiol-cypionate  Estrogen  200.00000
17               raloxifene  Estrogen    0.01710
18                masitinib      FGFR    2.50000
19                ponatinib      FGFR    0.00660
20                sorafenib      FGFR    1.74000
21               KU-0063794      PI3K    0.03000
22                   PI-103      PI3K    0.09000
23                  AMG-232       p53    0.00180
24                  HLI-373       p53    3.00000
25                 nutlin-3       p53    0.54000
26                     RITA       p53    9.75000
27                SAR405838       p53    0.02640
28                AS-605240      PI3K    0.15950
29                 GDC-0349      PI3K    3.00000
30                LY-294002      PI3K    1.36000
31               NVP-BEZ235      PI3K    1.71000
32                taselisib      PI3K    0.00262
33             daunorubicin     TOP2A    0.02000
34               epirubicin     TOP2A    0.17000
35               idarubicin     TOP2A    0.00990
36             mitoxantrone     TOP2A    5.30000
37                 AZD-8055      PI3K    0.27430
38               serdemetan       p53   62.15000

In [6]:
modules = IC50_df.Module.unique().tolist()

print('IC50_df  modules list: ', len(modules), modules)
print()
print('Selected modules list: ', len(selected_modules), selected_modules)

### CHECK
print()
print('CHECK: ', len(selected_modules),'=?', len(modules))

n_modules = len(modules)


IC50_df  modules list:  10 ['Androgen', 'CDK1', 'CDK2', 'CDK4_6', 'EGFR', 'Estrogen', 'FGFR', 'PI3K', 'p53', 'TOP2A']

Selected modules list:  10 ['CDK1', 'CDK2', 'CDK4_6', 'EGFR', 'PI3K', 'FGFR', 'TOP2A', 'p53', 'Estrogen', 'Androgen']

CHECK:  10 =? 10


In [7]:
drugs = IC50_df.Drug.tolist()
print(len(drugs), ' - Size after reading')

# remove duplicates
drugs = IC50_df.Drug.unique().tolist()
print(len(drugs), ' - Size after remove duplicates')

# remove whitespaces in drugs' names (necessary for some)
drugs = [d.strip() for d in drugs]

# remove duplicates after remove whitespaces
drugs = list(set(drugs))
print(len(drugs), ' - Size after remove duplicates without whitespaces')

print('Drugs list: ', len(drugs), drugs)

n_drugs = len(drugs)

39  - Size after reading
39  - Size after remove duplicates
39  - Size after remove duplicates without whitespaces
Drugs list:  39 ['testosterone-enanthate', 'ponatinib', 'estradiol-cypionate', 'oxandrolone', 'PI-103', 'afatinib', 'roscovitine', 'flufenamic-acid', 'testosterone-propionate', 'SAR405838', 'gefitinib', 'RITA', 'sorafenib', 'AZD-8055', 'HLI-373', 'palbociclib', 'masitinib', 'erlotinib', 'JNJ-7706621', 'NVP-BEZ235', 'serdemetan', 'taselisib', 'vandetanib', 'PHA-793887', 'idarubicin', 'GDC-0349', 'AMG-232', 'AS-605240', 'dienestrol', 'daunorubicin', 'raloxifene', 'mitoxantrone', 'lapatinib', 'KU-0063794', 'nutlin-3', 'alvocidib', 'epirubicin', 'nandrolone', 'LY-294002']


## L1000 meta data

Get sig_id for selected drugs.

In [8]:
sig_info_df = pd.read_excel(os.path.join(data_dir, f"sig_info_2020_{cell_line}.xlsx"), index_col = 0)

display(sig_info_df)

cell    plate  time              level_3_samples  \
level_5_sig_id                                                           
ASG002_BC3C_24H:A03   BC3C   ASG002  24 h   ASG002_BC3C_24H_X1_B35:A03   
ASG002_BC3C_24H:A04   BC3C   ASG002  24 h   ASG002_BC3C_24H_X1_B35:A04   
ASG002_BC3C_24H:A05   BC3C   ASG002  24 h   ASG002_BC3C_24H_X1_B35:A05   
ASG002_BC3C_24H:A06   BC3C   ASG002  24 h   ASG002_BC3C_24H_X1_B35:A06   
ASG002_BC3C_24H:J13   BC3C   ASG002  24 h   ASG002_BC3C_24H_X1_B35:J13   
...                    ...      ...   ...                          ...   
MOAR012_BC3C_24H:P20  BC3C  MOAR012  24 h  MOAR012_BC3C_24H_X1_B36:P20   
MOAR012_BC3C_24H:P21  BC3C  MOAR012  24 h  MOAR012_BC3C_24H_X1_B36:P21   
MOAR012_BC3C_24H:P22  BC3C  MOAR012  24 h  MOAR012_BC3C_24H_X1_B36:P22   
MOAR012_BC3C_24H:P23  BC3C  MOAR012  24 h  MOAR012_BC3C_24H_X1_B36:P23   
MOAR012_BC3C_24H:P24  BC3C  MOAR012  24 h  MOAR012_BC3C_24H_X1_B36:P24   

                      samples_number    pert_type    pert_drug  \
level_5_sig_id                                                   
ASG002_BC3C_24H:A03                1  ctl_vehicle         DMSO   
ASG002_BC3C_24H:A04                1  ctl_vehicle         DMSO   
ASG002_BC3C_24H:A05                1  ctl_vehicle         DMSO   
ASG002_BC3C_24H:A06                1  ctl_vehicle         DMSO   
ASG002_BC3C_24H:J13                1  ctl_vehicle         DMSO   
...                              ...          ...          ...   
MOAR012_BC3C_24H:P20               1       trt_cp  BAY-61-3606   
MOAR012_BC3C_24H:P21               1       trt_cp  BAY-61-3606   
MOAR012_BC3C_24H:P22               1       trt_cp   ethaverine   
MOAR012_BC3C_24H:P23               1       trt_cp   ethaverine   
MOAR012_BC3C_24H:P24               1       trt_cp   ethaverine   

                             targets  targets_number     dose  dose_float  
level_5_sig_id                                                             
ASG002_BC3C_24H:A03   DMSO_No_target               0     0 uM        0.00  
ASG002_BC3C_24H:A04   DMSO_No_target               0     0 uM        0.00  
ASG002_BC3C_24H:A05   DMSO_No_target               0     0 uM        0.00  
ASG002_BC3C_24H:A06   DMSO_No_target               0     0 uM        0.00  
ASG002_BC3C_24H:J13   DMSO_No_target               0     0 uM        0.00  
...                              ...             ...      ...         ...  
MOAR012_BC3C_24H:P20             NaN               0  3.33 uM        3.33  
MOAR012_BC3C_24H:P21             NaN               0  1.11 uM        1.11  
MOAR012_BC3C_24H:P22             NaN               0    10 uM       10.00  
MOAR012_BC3C_24H:P23             NaN               0  3.33 uM        3.33  
MOAR012_BC3C_24H:P24             NaN               0  1.11 uM        1.11  

[1971 rows x 11 columns]

In [54]:
# now filtering so only the required drugs are present
sig_info_df = sig_info_df.loc[sig_info_df.pert_drug.isin(drugs)]

# here's what we have now
display(sig_info_df)

cell    plate  time               level_3_samples  \
level_5_sig_id                                                              
ASG002_CAL29_24H:A10   CAL29   ASG002  24 h   ASG002_CAL29_24H_X1_B35:A10   
ASG002_CAL29_24H:A11   CAL29   ASG002  24 h   ASG002_CAL29_24H_X1_B35:A11   
ASG002_CAL29_24H:A12   CAL29   ASG002  24 h   ASG002_CAL29_24H_X1_B35:A12   
ASG002_CAL29_24H:A19   CAL29   ASG002  24 h   ASG002_CAL29_24H_X1_B35:A19   
ASG002_CAL29_24H:A20   CAL29   ASG002  24 h   ASG002_CAL29_24H_X1_B35:A20   
...                      ...      ...   ...                           ...   
MOAR011_CAL29_24H:C11  CAL29  MOAR011  24 h  MOAR011_CAL29_24H_X1_B35:C11   
MOAR011_CAL29_24H:C12  CAL29  MOAR011  24 h  MOAR011_CAL29_24H_X1_B35:C12   
MOAR011_CAL29_24H:J10  CAL29  MOAR011  24 h  MOAR011_CAL29_24H_X1_B35:J10   
MOAR011_CAL29_24H:J11  CAL29  MOAR011  24 h  MOAR011_CAL29_24H_X1_B35:J11   
MOAR011_CAL29_24H:J12  CAL29  MOAR011  24 h  MOAR011_CAL29_24H_X1_B35:J12   

                       samples_number pert_type               pert_drug  \
level_5_sig_id                                                            
ASG002_CAL29_24H:A10                1    trt_cp               taselisib   
ASG002_CAL29_24H:A11                1    trt_cp               taselisib   
ASG002_CAL29_24H:A12                1    trt_cp               taselisib   
ASG002_CAL29_24H:A19                1    trt_cp               AS-605240   
ASG002_CAL29_24H:A20                1    trt_cp               AS-605240   
...                               ...       ...                     ...   
MOAR011_CAL29_24H:C11               1    trt_cp  testosterone-enanthate   
MOAR011_CAL29_24H:C12               1    trt_cp  testosterone-enanthate   
MOAR011_CAL29_24H:J10               1    trt_cp              nandrolone   
MOAR011_CAL29_24H:J11               1    trt_cp              nandrolone   
MOAR011_CAL29_24H:J12               1    trt_cp              nandrolone   

                                              targets  targets_number  \
level_5_sig_id                                                          
ASG002_CAL29_24H:A10                           PIK3CA               1   
ASG002_CAL29_24H:A11                           PIK3CA               1   
ASG002_CAL29_24H:A12                           PIK3CA               1   
ASG002_CAL29_24H:A19                           PIK3CG               1   
ASG002_CAL29_24H:A20                           PIK3CG               1   
...                                               ...             ...   
MOAR011_CAL29_24H:C11                              AR               1   
MOAR011_CAL29_24H:C12                              AR               1   
MOAR011_CAL29_24H:J10  AR, CYP19A1, MAOA, MAOB, NISCH               5   
MOAR011_CAL29_24H:J11  AR, CYP19A1, MAOA, MAOB, NISCH               5   
MOAR011_CAL29_24H:J12  AR, CYP19A1, MAOA, MAOB, NISCH               5   

                          dose  dose_float  
level_5_sig_id                              
ASG002_CAL29_24H:A10     10 uM       10.00  
ASG002_CAL29_24H:A11   1.11 uM        1.11  
ASG002_CAL29_24H:A12   0.12 uM        0.12  
ASG002_CAL29_24H:A19     10 uM       10.00  
ASG002_CAL29_24H:A20   1.11 uM        1.11  
...                        ...         ...  
MOAR011_CAL29_24H:C11  3.33 uM        3.33  
MOAR011_CAL29_24H:C12  1.11 uM        1.11  
MOAR011_CAL29_24H:J10    10 uM       10.00  
MOAR011_CAL29_24H:J11  3.33 uM        3.33  
MOAR011_CAL29_24H:J12  1.11 uM        1.11  

[97 rows x 11 columns]

Manually remove  few inhibition from data set, since it does differ from the other data points.

inhib_to_filter = "PF-03758309"
dose_to_filter = 10

id_to_filter = sig_info_df[
    np.logical_and(
        sig_info_df.pert_drug == inhib_to_filter,
        sig_info_df.dose_float == dose_to_filter,
    )
].index.values

print(id_to_filter)

sig_info_df = sig_info_df[~sig_info_df.index.isin(id_to_filter)]
display(sig_info_df)

inhib_to_filter = "roscovitine"
dose_to_filter = 3.33

id_to_filter = sig_info_df[
    np.logical_and(
        sig_info_df.pert_drug == inhib_to_filter,
        sig_info_df.dose_float == dose_to_filter,
    )
].index.values

print(id_to_filter)

sig_info_df = sig_info_df[~sig_info_df.index.isin(id_to_filter)]
display(sig_info_df)

In [9]:
exp_ids = sig_info_df.index.unique().tolist()
print('Experiments ids list: ', len(exp_ids))

n_experiments = len(exp_ids)

Experiments ids list:  1971


Confirm data by checking the drugs of interest against the filtered L1000 meta data.

In [10]:
print(f"Number of drugs of interest:\t{len(drugs)}")
#print(f'Number of drugs in L1000 data:\t{len(sig_info_df.value_counts("drugs"))}')

#sig_info_df.value_counts("drugs")

Number of drugs of interest:	39


## L1000 data

In [11]:
Data_norm_df = pd.read_excel(os.path.join(data_dir, f"Data_norm_2020_{cell_line}.xlsx"), index_col = 0)
display(Data_norm_df)

AARS     ABCB6     ABCC5     ABCF1     ABCF3  \
ASG002_BC3C_24H:A03  -0.191254 -0.055246  0.039596 -0.256266 -0.040419   
ASG002_BC3C_24H:A04  -0.265754 -0.317496  0.118696 -0.136665 -0.301569   
ASG002_BC3C_24H:A05  -0.181954 -0.081597 -0.210304  1.559535 -0.019019   
ASG002_BC3C_24H:A06   0.033446  0.042404 -0.150154 -0.093165  0.053180   
ASG002_BC3C_24H:J13   0.204446  0.180704  0.089096 -0.054666  0.053381   
...                        ...       ...       ...       ...       ...   
MOAR012_BC3C_24H:P20  0.647151  0.211700 -0.979200  0.597350 -0.375751   
MOAR012_BC3C_24H:P21  0.171800  0.046300 -0.145550 -0.295150  0.030849   
MOAR012_BC3C_24H:P22  0.648700  0.058749 -0.031700  0.408249 -0.753950   
MOAR012_BC3C_24H:P23  0.090499 -0.469300 -0.611800  0.873550 -0.788450   
MOAR012_BC3C_24H:P24  0.204801 -1.388700 -0.283400  0.744750 -1.431150   

                         ABHD4     ABHD6      ABL1     ACAA1     ACAT2  ...  \
ASG002_BC3C_24H:A03  -0.590523 -0.159396 -0.074319  0.457981  0.409608  ...   
ASG002_BC3C_24H:A04  -0.403023  0.124804 -0.036470  0.311931  0.660457  ...   
ASG002_BC3C_24H:A05  -0.457423  0.071404  0.074080 -0.356119  0.498808  ...   
ASG002_BC3C_24H:A06  -0.053823  0.087704  0.167681 -0.601569  0.383308  ...   
ASG002_BC3C_24H:J13   0.044877 -0.277396 -0.157419  0.535681 -3.933493  ...   
...                        ...       ...       ...       ...       ...  ...   
MOAR012_BC3C_24H:P20  0.388300  0.394524  0.120151 -0.166775 -1.129125  ...   
MOAR012_BC3C_24H:P21  0.420951  0.222075  0.179800  0.274724 -0.423975  ...   
MOAR012_BC3C_24H:P22  0.332200 -0.357525 -0.107650 -0.213575  0.074225  ...   
MOAR012_BC3C_24H:P23 -0.097199 -0.366575 -0.490600 -0.624675 -0.009275  ...   
MOAR012_BC3C_24H:P24 -0.518100 -0.002174 -0.096200  1.198824  0.138525  ...   

                         ZMIZ1     ZMYM2    ZNF131    ZNF274    ZNF318  \
ASG002_BC3C_24H:A03   0.543203  0.494266 -0.011923 -0.225931  0.285054   
ASG002_BC3C_24H:A04  -0.565096 -0.088634  0.122977 -0.047931  0.141804   
ASG002_BC3C_24H:A05   0.226104 -0.228034 -0.121023 -0.075331 -0.133146   
ASG002_BC3C_24H:A06  -0.608596 -0.228835  0.072777  0.082970 -0.570996   
ASG002_BC3C_24H:J13  -0.318397  0.122265 -0.134323 -0.088931 -0.067996   
...                        ...       ...       ...       ...       ...   
MOAR012_BC3C_24H:P20 -1.598475 -0.552750  0.515151  0.120800  0.082675   
MOAR012_BC3C_24H:P21 -1.650575  0.203600 -0.003250 -0.064800 -0.037675   
MOAR012_BC3C_24H:P22  0.127625  0.031600  0.103250 -0.249600  0.046375   
MOAR012_BC3C_24H:P23  0.054676 -0.596050  0.084600  0.444700  0.431375   
MOAR012_BC3C_24H:P24 -0.123474 -0.486500  0.049901 -0.598100 -0.259925   

                        ZNF395    ZNF451    ZNF586    ZNF589      ZW10  
ASG002_BC3C_24H:A03  -0.775246  0.166031 -0.024873  0.238723  0.284204  
ASG002_BC3C_24H:A04   0.129054 -0.028819 -0.028773 -0.253627 -0.752646  
ASG002_BC3C_24H:A05   0.355054  0.022831 -0.084073  0.283123 -0.894896  
ASG002_BC3C_24H:A06   2.847754 -0.211670 -0.067273  0.081723  0.338704  
ASG002_BC3C_24H:J13  -0.515847 -0.005069  0.067527  0.002223  0.204904  
...                        ...       ...       ...       ...       ...  
MOAR012_BC3C_24H:P20  0.529700  0.383225 -0.207225  2.268450 -1.248500  
MOAR012_BC3C_24H:P21  0.076499  0.201825  0.416875  0.287450 -0.971700  
MOAR012_BC3C_24H:P22  1.486200  0.440325  0.090075 -0.031650 -0.944300  
MOAR012_BC3C_24H:P23 -0.921501  0.044926  0.716076 -0.000900 -1.106700  
MOAR012_BC3C_24H:P24  0.088650 -0.242075  0.195575 -0.038300 -1.149900  

[1971 rows x 978 columns]

In [12]:
Data_norm_df = Data_norm_df[Data_norm_df.index.isin(exp_ids)]

# arrange experiments in same order as in list
Data_norm_df["sort_col"] = Data_norm_df.index.map({val: i for i, val in enumerate(exp_ids)})
Data_norm_df = Data_norm_df.sort_values("sort_col")
Data_norm_df = Data_norm_df.drop("sort_col", axis = 1)

# transpose
Data_norm_df = Data_norm_df.T

display(Data_norm_df)

ASG002_BC3C_24H:A03  ASG002_BC3C_24H:A04  ASG002_BC3C_24H:A05  \
AARS              -0.191254            -0.265754            -0.181954   
ABCB6             -0.055246            -0.317496            -0.081597   
ABCC5              0.039596             0.118696            -0.210304   
ABCF1             -0.256266            -0.136665             1.559535   
ABCF3             -0.040419            -0.301569            -0.019019   
...                     ...                  ...                  ...   
ZNF395            -0.775246             0.129054             0.355054   
ZNF451             0.166031            -0.028819             0.022831   
ZNF586            -0.024873            -0.028773            -0.084073   
ZNF589             0.238723            -0.253627             0.283123   
ZW10               0.284204            -0.752646            -0.894896   

        ASG002_BC3C_24H:A06  ASG002_BC3C_24H:J13  ASG002_BC3C_24H:J14  \
AARS               0.033446             0.204446             0.408096   
ABCB6              0.042404             0.180704             0.170904   
ABCC5             -0.150154             0.089096             0.326696   
ABCF1             -0.093165            -0.054666            -0.468565   
ABCF3              0.053180             0.053381             0.304181   
...                     ...                  ...                  ...   
ZNF395             2.847754            -0.515847             1.230454   
ZNF451            -0.211670            -0.005069             0.182881   
ZNF586            -0.067273             0.067527             0.023977   
ZNF589             0.081723             0.002223            -0.123027   
ZW10               0.338704             0.204904             0.542403   

        ASG002_BC3C_24H:J15  ASG002_BC3C_24H:J16  ASG002_BC3C_24H:J17  \
AARS               0.343246             0.294746             0.097946   
ABCB6              0.264104             0.679953             0.212804   
ABCC5              0.346296             0.872796            -0.025904   
ABCF1             -0.439466            -0.008215            -0.174766   
ABCF3              0.209981            -0.061119             0.364280   
...                     ...                  ...                  ...   
ZNF395            -2.410447            -1.559446            -0.790896   
ZNF451             0.017630             0.327481            -0.145819   
ZNF586             0.063227             0.108677            -0.018573   
ZNF589            -0.165777             0.007073             0.153373   
ZW10              -0.689296             0.414804             0.261904   

        ASG002_BC3C_24H:J18  ...  MOAR012_BC3C_24H:P15  MOAR012_BC3C_24H:P16  \
AARS               0.174596  ...              0.359300              0.235600   
ABCB6              0.060254  ...              0.111150             -0.022400   
ABCC5              0.393696  ...             -1.173200             -0.649050   
ABCF1              0.021435  ...              1.054650              0.868550   
ABCF3              0.550581  ...             -1.016601             -0.464351   
...                     ...  ...                   ...                   ...   
ZNF395            -1.058396  ...              1.914200              2.336899   
ZNF451            -0.140719  ...              0.354325              0.487625   
ZNF586             0.139427  ...              0.188375              0.048375   
ZNF589            -0.119577  ...             -0.677950             -0.111700   
ZW10              -0.885796  ...             -1.325700             -0.622000   

        MOAR012_BC3C_24H:P17  MOAR012_BC3C_24H:P18  MOAR012_BC3C_24H:P19  \
AARS               -0.160699              0.351300              0.900100   
ABCB6              -0.814200              0.043200              0.710800   
ABCC5               2.041700             -0.674200             -0.611500   
ABCF1               1.379700              0.669800              1.291850   
ABCF3              -1.356300        

In [13]:
genes = Data_norm_df.index.tolist()
print('Landmark genes list: ', len(genes), genes)

n_genes = len(genes)

Landmark genes list:  978 ['AARS', 'ABCB6', 'ABCC5', 'ABCF1', 'ABCF3', 'ABHD4', 'ABHD6', 'ABL1', 'ACAA1', 'ACAT2', 'ACBD3', 'ACD', 'ACLY', 'ACOT9', 'ADAM10', 'ADAT1', 'ADGRE5', 'ADGRG1', 'ADH5', 'ADI1', 'ADO', 'ADRB2', 'AGL', 'AKAP8', 'AKAP8L', 'AKR7A2', 'AKT1', 'ALAS1', 'ALDH7A1', 'ALDOA', 'ALDOC', 'AMDHD2', 'ANKRD10', 'ANO10', 'ANXA7', 'APBB2', 'APOE', 'APP', 'APPBP2', 'ARFIP2', 'ARHGAP1', 'ARHGEF12', 'ARHGEF2', 'ARID4B', 'ARID5B', 'ARL4C', 'ARNT2', 'ARPP19', 'ASAH1', 'ASCC3', 'ATF1', 'ATF5', 'ATF6', 'ATG3', 'ATMIN', 'ATP11B', 'ATP1B1', 'ATP2C1', 'ATP6V0B', 'ATP6V1D', 'AURKA', 'AURKB', 'AXIN1', 'B4GAT1', 'BACE2', 'BAD', 'BAG3', 'BAMBI', 'BAX', 'BCL2', 'BCL7B', 'BDH1', 'BECN1', 'BHLHE40', 'BID', 'BIRC2', 'BIRC5', 'BLCAP', 'BLMH', 'BLVRA', 'BMP4', 'BNIP3', 'BNIP3L', 'BPHL', 'BRCA1', 'BTK', 'BUB1B', 'BZW2', 'C2CD2', 'C2CD2L', 'C2CD5', 'C5', 'CAB39', 'CALM3', 'CALU', 'CAMSAP2', 'CANT1', 'CAPN1', 'CARMIL1', 'CASC3', 'CASK', 'CASP10', 'CASP2', 'CASP3', 'CASP7', 'CAST', 'CAT', 'CBLB', 'CBR1

## Inhibitor concentrations, IC50, and perturbation matrices

In [14]:
inhib_conc_matrix = np.zeros((n_modules, n_experiments))
ic50_matrix = np.ones((n_modules, n_experiments))
gamma_matrix = np.zeros((n_modules, n_experiments))

In [15]:

for i, module in enumerate(modules):
    drugs_for_module = IC50_df.Drug[IC50_df.Module == module].tolist()
    for drug in drugs_for_module:
        # get IC50 for this drug
        ic50 = IC50_df.IC50[IC50_df.Drug == drug].values
#       gamma = IC50_df.Gamma[IC50_df.Drug == drug].values
        print(drug, ic50)
        assert ic50.size == 1
#       assert gamma.size == 1
        # get experiments with this drug
        exp_with_drug = sig_info_df.index[sig_info_df.pert_drug == drug].tolist()
        print(exp_with_drug) 
        for exp_id in exp_with_drug:
            j = exp_ids.index(exp_id)
            print(j)
            # extract inhibitor concentration
            inhib_conc = sig_info_df.dose_float[sig_info_df.index == exp_id].values
            assert inhib_conc.size == 1
            # insert values in matrices
            inhib_conc_matrix[i, j] = inhib_conc.item()
            ic50_matrix[i, j] = ic50.item()
#           gamma_matrix[i, j] = gamma.item()


flufenamic-acid [3.]
['MOAR008_BC3C_24H:L01', 'MOAR008_BC3C_24H:L02', 'MOAR008_BC3C_24H:L03']
623
624
625
nandrolone [9.]
['MOAR011_BC3C_24H:J10']
1555
oxandrolone [190.3]
['MOAR011_BC3C_24H:C01', 'MOAR011_BC3C_24H:C02', 'MOAR011_BC3C_24H:C03']
1475
1476
1477
testosterone-enanthate [200.]
['MOAR011_BC3C_24H:C10', 'MOAR011_BC3C_24H:C11']
1484
1485
testosterone-propionate [124.]
['MOAR010_BC3C_24H:L19', 'MOAR010_BC3C_24H:L20', 'MOAR010_BC3C_24H:L21']
1360
1361
1362
JNJ-7706621 [0.027]
['ASG002_BC3C_24H:N13', 'ASG002_BC3C_24H:N14', 'ASG002_BC3C_24H:N15']
323
324
325
PHA-793887 [0.18]
['ASG002_BC3C_24H:L01', 'ASG002_BC3C_24H:L02', 'ASG002_BC3C_24H:L03']
264
265
266
roscovitine [2.]
['ASG002_BC3C_24H:E22', 'ASG002_BC3C_24H:E23', 'ASG002_BC3C_24H:E24']
132
133
134
alvocidib [0.12]
['ASG002_BC3C_24H:F04', 'ASG002_BC3C_24H:F05', 'ASG002_BC3C_24H:F06']
138
139
140
palbociclib [0.045]
['ASG002_BC3C_24H:P16', 'ASG002_BC3C_24H:P17', 'ASG002_BC3C_24H:P18']
371
372
373
afatinib [0.03]
['ASG002_BC3C_

In [16]:
# transform matrices into pandas dfs for export with row and column names
inhib_conc_df = pd.DataFrame(inhib_conc_matrix, index = modules, columns = exp_ids)
ic50_df = pd.DataFrame(ic50_matrix, index = modules, columns = exp_ids)
# gamma_df = pd.DataFrame(gamma_matrix, index = modules, columns = exp_ids)

# create binary perturbation matrix
pert_df = pd.DataFrame(
    np.where(inhib_conc_matrix != 0, 1, 0),
    index = inhib_conc_df.index,
    columns = inhib_conc_df.columns,
)

In [17]:
display(ic50_df)
# display(gamma_df)
display(inhib_conc_df)
display(pert_df)

ASG002_BC3C_24H:A03  ASG002_BC3C_24H:A04  ASG002_BC3C_24H:A05  \
Androgen                  1.0                  1.0                  1.0   
CDK1                      1.0                  1.0                  1.0   
CDK2                      1.0                  1.0                  1.0   
CDK4_6                    1.0                  1.0                  1.0   
EGFR                      1.0                  1.0                  1.0   
Estrogen                  1.0                  1.0                  1.0   
FGFR                      1.0                  1.0                  1.0   
PI3K                      1.0                  1.0                  1.0   
p53                       1.0                  1.0                  1.0   
TOP2A                     1.0                  1.0                  1.0   

          ASG002_BC3C_24H:A06  ASG002_BC3C_24H:J13  ASG002_BC3C_24H:J14  \
Androgen                  1.0                  1.0                  1.0   
CDK1                      1.0                  1.0                  1.0   
CDK2                      1.0                  1.0                  1.0   
CDK4_6                    1.0                  1.0                  1.0   
EGFR                      1.0                  1.0                  1.0   
Estrogen                  1.0                  1.0                  1.0   
FGFR                      1.0                  1.0                  1.0   
PI3K                      1.0                  1.0                  1.0   
p53                       1.0                  1.0                  1.0   
TOP2A                     1.0                  1.0                  1.0   

          ASG002_BC3C_24H:J15  ASG002_BC3C_24H:J16  ASG002_BC3C_24H:J17  \
Androgen                  1.0                  1.0                  1.0   
CDK1                      1.0                  1.0                  1.0   
CDK2                      1.0                  1.0                  1.0   
CDK4_6                    1.0                  1.0                  1.0   
EGFR                      1.0                  1.0                  1.0   
Estrogen                  1.0                  1.0                  1.0   
FGFR                      1.0                  1.0                  1.0   
PI3K                      1.0                  1.0                  1.0   
p53                       1.0                  1.0                  1.0   
TOP2A                     1.0                  1.0                  1.0   

          ASG002_BC3C_24H:J18  ...  MOAR012_BC3C_24H:P15  \
Androgen                  1.0  ...                   1.0   
CDK1                      1.0  ...                   1.0   
CDK2                      1.0  ...                   1.0   
CDK4_6                    1.0  ...                   1.0   
EGFR                      1.0  ...                   1.0   
Estrogen                  1.0  ...                   1.0   
FGFR                      1.0  ...                   1.0   
PI3K                      1.0  ...                   1.0   
p53                       1.0  ...                   1.0   
TOP2A                     1.0  ...                   1.0   

          MOAR012_BC3C_24H:P16  MOAR012_BC3C_24H:P17  MOAR012_BC3C_24H:P18  \
Androgen                   1.0                   1.0                   1.0   
CDK1                       1.0                   1.0                   1.0   
CDK2                       1.0                   1.0                   1.0   
CDK4_6                     1.0                   1.0                   1.0   
EGFR                       1.0                   1.0                   1.0   
Estrogen                   1.0                   1.0                   1.0   
FGFR                       1.0                   1.0                   1.0   
PI3K                       1.0                   1.0                   1.0   
p53                        1.0                   1.0                   1.0   
TOP2A                      1.0                   1.0                   1.0   

          MO

ASG002_BC3C_24H:A03  ASG002_BC3C_24H:A04  ASG002_BC3C_24H:A05  \
Androgen                  0.0                  0.0                  0.0   
CDK1                      0.0                  0.0                  0.0   
CDK2                      0.0                  0.0                  0.0   
CDK4_6                    0.0                  0.0                  0.0   
EGFR                      0.0                  0.0                  0.0   
Estrogen                  0.0                  0.0                  0.0   
FGFR                      0.0                  0.0                  0.0   
PI3K                      0.0                  0.0                  0.0   
p53                       0.0                  0.0                  0.0   
TOP2A                     0.0                  0.0                  0.0   

          ASG002_BC3C_24H:A06  ASG002_BC3C_24H:J13  ASG002_BC3C_24H:J14  \
Androgen                  0.0                  0.0                  0.0   
CDK1                      0.0                  0.0                  0.0   
CDK2                      0.0                  0.0                  0.0   
CDK4_6                    0.0                  0.0                  0.0   
EGFR                      0.0                  0.0                  0.0   
Estrogen                  0.0                  0.0                  0.0   
FGFR                      0.0                  0.0                  0.0   
PI3K                      0.0                  0.0                  0.0   
p53                       0.0                  0.0                  0.0   
TOP2A                     0.0                  0.0                  0.0   

          ASG002_BC3C_24H:J15  ASG002_BC3C_24H:J16  ASG002_BC3C_24H:J17  \
Androgen                  0.0                  0.0                  0.0   
CDK1                      0.0                  0.0                  0.0   
CDK2                      0.0                  0.0                  0.0   
CDK4_6                    0.0                  0.0                  0.0   
EGFR                      0.0                  0.0                  0.0   
Estrogen                  0.0                  0.0                  0.0   
FGFR                      0.0                  0.0                  0.0   
PI3K                      0.0                  0.0                  0.0   
p53                       0.0                  0.0                  0.0   
TOP2A                     0.0                  0.0                  0.0   

          ASG002_BC3C_24H:J18  ...  MOAR012_BC3C_24H:P15  \
Androgen                  0.0  ...                   0.0   
CDK1                      0.0  ...                   0.0   
CDK2                      0.0  ...                   0.0   
CDK4_6                    0.0  ...                   0.0   
EGFR                      0.0  ...                   0.0   
Estrogen                  0.0  ...                   0.0   
FGFR                      0.0  ...                   0.0   
PI3K                      0.0  ...                   0.0   
p53                       0.0  ...                   0.0   
TOP2A                     0.0  ...                   0.0   

          MOAR012_BC3C_24H:P16  MOAR012_BC3C_24H:P17  MOAR012_BC3C_24H:P18  \
Androgen                   0.0                   0.0                   0.0   
CDK1                       0.0                   0.0                   0.0   
CDK2                       0.0                   0.0                   0.0   
CDK4_6                     0.0                   0.0                   0.0   
EGFR                       0.0                   0.0                   0.0   
Estrogen                   0.0                   0.0                   0.0   
FGFR                       0.0                   0.0                   0.0   
PI3K                       0.0                   0.0                   0.0   
p53                        0.0                   0.0                   0.0   
TOP2A                      0.0                   0.0                   0.0   

          MO

ASG002_BC3C_24H:A03  ASG002_BC3C_24H:A04  ASG002_BC3C_24H:A05  \
Androgen                    0                    0                    0   
CDK1                        0                    0                    0   
CDK2                        0                    0                    0   
CDK4_6                      0                    0                    0   
EGFR                        0                    0                    0   
Estrogen                    0                    0                    0   
FGFR                        0                    0                    0   
PI3K                        0                    0                    0   
p53                         0                    0                    0   
TOP2A                       0                    0                    0   

          ASG002_BC3C_24H:A06  ASG002_BC3C_24H:J13  ASG002_BC3C_24H:J14  \
Androgen                    0                    0                    0   
CDK1                        0                    0                    0   
CDK2                        0                    0                    0   
CDK4_6                      0                    0                    0   
EGFR                        0                    0                    0   
Estrogen                    0                    0                    0   
FGFR                        0                    0                    0   
PI3K                        0                    0                    0   
p53                         0                    0                    0   
TOP2A                       0                    0                    0   

          ASG002_BC3C_24H:J15  ASG002_BC3C_24H:J16  ASG002_BC3C_24H:J17  \
Androgen                    0                    0                    0   
CDK1                        0                    0                    0   
CDK2                        0                    0                    0   
CDK4_6                      0                    0                    0   
EGFR                        0                    0                    0   
Estrogen                    0                    0                    0   
FGFR                        0                    0                    0   
PI3K                        0                    0                    0   
p53                         0                    0                    0   
TOP2A                       0                    0                    0   

          ASG002_BC3C_24H:J18  ...  MOAR012_BC3C_24H:P15  \
Androgen                    0  ...                     0   
CDK1                        0  ...                     0   
CDK2                        0  ...                     0   
CDK4_6                      0  ...                     0   
EGFR                        0  ...                     0   
Estrogen                    0  ...                     0   
FGFR                        0  ...                     0   
PI3K                        0  ...                     0   
p53                         0  ...                     0   
TOP2A                       0  ...                     0   

          MOAR012_BC3C_24H:P16  MOAR012_BC3C_24H:P17  MOAR012_BC3C_24H:P18  \
Androgen                     0                     0                     0   
CDK1                         0                     0                     0   
CDK2                         0                     0                     0   
CDK4_6                       0                     0                     0   
EGFR                         0                     0                     0   
Estrogen                     0                     0                     0   
FGFR                         0                     0                     0   
PI3K                         0                     0                     0   
p53                          0                     0                     0   
TOP2A                        0                     0                     0   

          MO

## Global responses for DPD modules

According to our discussion, $R$ for DPD vectors can not be calculated with the same formula as for pathway activities. Instead we are assuming:

\begin{equation}
R_{DPD},j = DPD = STV_{DPD} \cdot Data_j
\end{equation}

In [18]:
# load STV data frame
STVs = pd.read_excel(os.path.join(info_dir, "ALL_DATA_2020_Jing.xlsx"), sheet_name = "STV", index_col = 0)
STV_df = pd.DataFrame(np.zeros((len(Data_norm_df.index), 3)), index = Data_norm_df.index, columns = STVs.columns)
STV_df.loc[STVs.index] = STVs

display(STV_df)

blca_invasiveness  blca_oncogenesis  blca_survival
AARS             0.016052          0.078487       0.000000
ABCB6            0.001612          0.027125      -2.744978
ABCC5           -0.016742          0.002121       0.000000
ABCF1           -0.038990         -0.028959      -0.555168
ABCF3            0.023288          0.036671       0.000000
...                   ...               ...            ...
ZNF395           0.000000          0.000000       0.000000
ZNF451           0.011450          0.013403       0.000000
ZNF586           0.000000          0.000000       0.000000
ZNF589           0.000000          0.000000       0.000000
ZW10            -0.042841          0.009722       0.000000

[978 rows x 3 columns]

In [19]:
# create empty DPD data frame
DPD_df = pd.DataFrame(
    np.zeros((len(Data_norm_df.columns), len(STV_df.columns))),
    index = Data_norm_df.columns,
    columns = STV_df.columns,
)

# populate
for exp_id in DPD_df.index:
    for state in STV_df.columns:
        DPD_df.loc[exp_id, state] = np.dot(Data_norm_df.T.loc[exp_id], STV_df.loc[:, state])

display(DPD_df)

blca_invasiveness  blca_oncogenesis  blca_survival
ASG002_BC3C_24H:A03            0.224169          0.302763      -2.305911
ASG002_BC3C_24H:A04            0.150151         -0.135368      -3.117183
ASG002_BC3C_24H:A05            0.167952         -0.014359      -2.260927
ASG002_BC3C_24H:A06            0.328579         -0.094210      -6.709615
ASG002_BC3C_24H:J13           -0.582389         -0.243571       0.612093
...                                 ...               ...            ...
MOAR012_BC3C_24H:P20           1.639600         -0.777955       3.452730
MOAR012_BC3C_24H:P21           0.843686         -0.200156       0.625054
MOAR012_BC3C_24H:P22           0.530906          0.472079      -1.294036
MOAR012_BC3C_24H:P23           0.684520         -0.052636      -7.776345
MOAR012_BC3C_24H:P24          -0.184652          0.325686     -22.252404

[1971 rows x 3 columns]

In [20]:
# transform to R global
R_global_DPD_df = DPD_df.T
display(R_global_DPD_df)

ASG002_BC3C_24H:A03  ASG002_BC3C_24H:A04  \
blca_invasiveness             0.224169             0.150151   
blca_oncogenesis              0.302763            -0.135368   
blca_survival                -2.305911            -3.117183   

                   ASG002_BC3C_24H:A05  ASG002_BC3C_24H:A06  \
blca_invasiveness             0.167952             0.328579   
blca_oncogenesis             -0.014359            -0.094210   
blca_survival                -2.260927            -6.709615   

                   ASG002_BC3C_24H:J13  ASG002_BC3C_24H:J14  \
blca_invasiveness            -0.582389             0.090671   
blca_oncogenesis             -0.243571             0.322869   
blca_survival                 0.612093             3.529396   

                   ASG002_BC3C_24H:J15  ASG002_BC3C_24H:J16  \
blca_invasiveness            -0.280056            -0.054662   
blca_oncogenesis             -0.202286             0.984030   
blca_survival                 3.110927             3.094277   

                   ASG002_BC3C_24H:J17  ASG002_BC3C_24H:J18  ...  \
blca_invasiveness            -0.301350            -0.032421  ...   
blca_oncogenesis              0.188662             0.383745  ...   
blca_survival                 3.581075             3.071644  ...   

                   MOAR012_BC3C_24H:P15  MOAR012_BC3C_24H:P16  \
blca_invasiveness              0.794299              0.649506   
blca_oncogenesis               0.437785             -0.296724   
blca_survival                  7.527461              0.666810   

                   MOAR012_BC3C_24H:P17  MOAR012_BC3C_24H:P18  \
blca_invasiveness             -0.550905              1.315166   
blca_oncogenesis              -0.957768             -0.311339   
blca_survival                  3.964668             -5.086449   

                   MOAR012_BC3C_24H:P19  MOAR012_BC3C_24H:P20  \
blca_invasiveness              0.853185              1.639600   
blca_oncogenesis               0.329503             -0.777955   
blca_survival                -11.084706              3.452730   

                   MOAR012_BC3C_24H:P21  MOAR012_BC3C_24H:P22  \
blca_invasiveness              0.843686              0.530906   
blca_oncogenesis              -0.200156              0.472079   
blca_survival                  0.625054             -1.294036   

                   MOAR012_BC3C_24H:P23  MOAR012_BC3C_24H:P24  
blca_invasiveness              0.684520             -0.184652  
blca_oncogenesis              -0.052636              0.325686  
blca_survival                 -7.776345            -22.252404  

[3 rows x 1971 columns]

## Save outputs

In [21]:
# save metadata as pickle
all_metadata = {
    "modules": modules,
    "n_modules": n_modules,
    "drugs": drugs,
    "n_drugs": n_drugs,
    "exp_ids": exp_ids,
    "n_experiments": n_experiments,
    "genes": genes,
    "n_genes": n_genes,
}

print(all_metadata)

with open(os.path.join(out_dir, "metadata.pickle"), "wb") as f:
    pickle.dump(all_metadata, f, protocol = pickle.HIGHEST_PROTOCOL)

{'modules': ['Androgen', 'CDK1', 'CDK2', 'CDK4_6', 'EGFR', 'Estrogen', 'FGFR', 'PI3K', 'p53', 'TOP2A'], 'n_modules': 10, 'drugs': ['testosterone-enanthate', 'ponatinib', 'estradiol-cypionate', 'oxandrolone', 'PI-103', 'afatinib', 'roscovitine', 'flufenamic-acid', 'testosterone-propionate', 'SAR405838', 'gefitinib', 'RITA', 'sorafenib', 'AZD-8055', 'HLI-373', 'palbociclib', 'masitinib', 'erlotinib', 'JNJ-7706621', 'NVP-BEZ235', 'serdemetan', 'taselisib', 'vandetanib', 'PHA-793887', 'idarubicin', 'GDC-0349', 'AMG-232', 'AS-605240', 'dienestrol', 'daunorubicin', 'raloxifene', 'mitoxantrone', 'lapatinib', 'KU-0063794', 'nutlin-3', 'alvocidib', 'epirubicin', 'nandrolone', 'LY-294002'], 'n_drugs': 39, 'exp_ids': ['ASG002_BC3C_24H:A03', 'ASG002_BC3C_24H:A04', 'ASG002_BC3C_24H:A05', 'ASG002_BC3C_24H:A06', 'ASG002_BC3C_24H:J13', 'ASG002_BC3C_24H:J14', 'ASG002_BC3C_24H:J15', 'ASG002_BC3C_24H:J16', 'ASG002_BC3C_24H:J17', 'ASG002_BC3C_24H:J18', 'ASG002_BC3C_24H:O04', 'ASG002_BC3C_24H:O05', 'ASG002

In [22]:
# save doses and perturbation matrix
inhib_conc_df.to_csv(os.path.join(out_dir, "inhib_conc_annotated.csv"))
ic50_df.to_csv(os.path.join(out_dir, "ic50_annotated.csv"))
# gamma_df.to_csv(os.path.join(out_dir, "gamma_annotated.csv"))
pert_df.to_csv(os.path.join(out_dir, "pert_annotated.csv"))

In [23]:
# save log fold change L1000 data
Data_norm_df.to_csv(os.path.join(out_dir, "L1000_Data_norm_data.csv"))

In [24]:
# save R_global for DPDs
R_global_DPD_df.to_csv(os.path.join(out_dir, "R_global_DPDonly_annotated.csv"))